Gastly crashes GoIV

1. Start service
2. Try to scan Gastly, 4 out of 5 times the IV button just disappears.
3. Go back to GoIV screen, click STOP button, the application closes with "Unfortunately, GoIV has stopped.".

Samsung Note 5, 6.0.1, GoIV 2.1.0. It didn't happen with the last version from the Play Store, it would just fail to recognize the CP when the smoke covered it.

Confirmed problem when screen scrolled up - IV button disappears, nothing happens and when GoIV is closed, following error occures:


> 08-23 13:01:18.568 31750-31750/com.kamron.pogoiv.nointernet E/AndroidRuntime: FATAL EXCEPTION: main
>                                                                               Process: com.kamron.pogoiv.nointernet, PID: 31750
>                                                                               java.lang.RuntimeException: Unable to stop service com.kamron.pogoiv.Pokefly@c5cb63d: java.lang.IllegalArgumentException: View=android.widget.ImageView{6a8b94 V.ED..... ......I. 0,0-0,0} not attached to window manager
>                                                                                   at android.app.ActivityThread.handleStopService(ActivityThread.java:3116)
>                                                                                   at android.app.ActivityThread.access$2300(ActivityThread.java:154)
>                                                                                   at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1494)
>                                                                                   at android.os.Handler.dispatchMessage(Handler.java:102)
>                                                                                   at android.os.Looper.loop(Looper.java:234)
>                                                                                   at android.app.ActivityThread.main(ActivityThread.java:5526)
>                                                                                   at java.lang.reflect.Method.invoke(Native Method)
>                                                                                   at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
>                                                                                   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
>                                                                                Caused by: java.lang.IllegalArgumentException: View=android.widget.ImageView{6a8b94 V.ED..... ......I. 0,0-0,0} not attached to window manager
>                                                                                   at android.view.WindowManagerGlobal.findViewLocked(WindowManagerGlobal.java:424)
>                                                                                   at android.view.WindowManagerGlobal.removeView(WindowManagerGlobal.java:350)
>                                                                                   at android.view.WindowManagerImpl.removeView(WindowManagerImpl.java:113)
>                                                                                   at com.kamron.pogoiv.Pokefly.onDestroy(Pokefly.java:224)
>                                                                                   at android.app.ActivityThread.handleStopService(ActivityThread.java:3097)
>                                                                                   at android.app.ActivityThread.access$2300(ActivityThread.java:154) 
>                                                                                   at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1494) 
>                                                                                   at android.os.Handler.dispatchMessage(Handler.java:102) 
>                                                                                   at android.os.Looper.loop(Looper.java:234) 
>                                                                                   at android.app.ActivityThread.main(ActivityThread.java:5526) 
>                                                                                   at java.lang.reflect.Method.invoke(Native Method) 

--- a/app/src/main/java/com/kamron/pogoiv/MainActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/MainActivity.java
@@ -630,6 +630,7 @@ public class MainActivity extends AppCompatActivity {
      * @param filePath The screenshot path if it is a file, used to delete once checked
      */
     private void scanPokemon(Bitmap pokemonImage, String filePath) {
+        //WARNING: this method *must* always send an intent at the end, no matter what, to avoid the application hanging.
         estimatedPokemonLevel = trainerLevel + 1.5;
 
         estimatedPokemonLevel= getPokemonLevel(pokemonImage);
@@ -706,9 +707,9 @@ public class MainActivity extends AppCompatActivity {
 
             Intent info = Pokefly.createInfoIntent(pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
             LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
-        }
-        else{
-            readyForNewScreenshot = true;
+        } else {
+            Intent info = Pokefly.createNoInfoIntent();
+            LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
         }
     }
 

--- a/app/src/main/java/com/kamron/pogoiv/MainActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/MainActivity.java
@@ -89,6 +89,9 @@ public class MainActivity extends AppCompatActivity {
 
     private static final String ACTION_RESET_SCREENSHOT = "reset-screenshot";
     private static final String ACTION_SCREENSHOT = "screenshot";
+    private static final String ACTION_PROCESS_BITMAP = "process-bitmap";
+
+    private static final String KEY_BITMAP = "bitmap";
 
     private MediaProjection mProjection;
     private ImageReader mImageReader;
@@ -132,6 +135,11 @@ public class MainActivity extends AppCompatActivity {
 
     public static Intent createResetScreenshotIntent() {
         return new Intent(ACTION_RESET_SCREENSHOT);
+    }
+    public static Intent createProcessBitmapIntent(Bitmap bitmap) {
+        Intent intent = new Intent(ACTION_PROCESS_BITMAP);
+        intent.putExtra(KEY_BITMAP, bitmap);
+        return intent;
     }
 
     @TargetApi(23)
@@ -262,9 +270,9 @@ public class MainActivity extends AppCompatActivity {
             candyOrder = 1;
         }
 
-        LocalBroadcastManager.getInstance(this).registerReceiver(resetScreenshot, new IntentFilter("reset-screenshot"));
-        LocalBroadcastManager.getInstance(this).registerReceiver(takeScreenshot, new IntentFilter("screenshot"));
-        LocalBroadcastManager.getInstance(this).registerReceiver(processBitmap, new IntentFilter("process-bitmap"));
+        LocalBroadcastManager.getInstance(this).registerReceiver(resetScreenshot, new IntentFilter(ACTION_RESET_SCREENSHOT));
+        LocalBroadcastManager.getInstance(this).registerReceiver(takeScreenshot, new IntentFilter(ACTION_SCREENSHOT));
+        LocalBroadcastManager.getInstance(this).registerReceiver(processBitmap, new IntentFilter(ACTION_PROCESS_BITMAP));
     }
 
     @Override
@@ -876,7 +884,7 @@ public class MainActivity extends AppCompatActivity {
         @Override
         public void onReceive(Context context, Intent intent) {
             if (readyForNewScreenshot) {
-                Bitmap bitmap = (Bitmap) intent.getParcelableExtra("bitmap");
+                Bitmap bitmap = (Bitmap) intent.getParcelableExtra(KEY_BITMAP);
                 scanPokemon(bitmap, "");
                 bitmap.recycle();
                 readyForNewScreenshot = false;

--- a/app/src/main/java/com/kamron/pogoiv/MainActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/MainActivity.java
@@ -109,7 +109,6 @@ public class MainActivity extends AppCompatActivity {
     private boolean readyForNewScreenshot = true;
 
     private int candyOrder = 0;
-    private double estimatedPokemonLevel;
     private boolean pokeFlyRunning = false;
     private int trainerLevel;
 
@@ -589,12 +588,14 @@ public class MainActivity extends AppCompatActivity {
         }
     }
 
+
     /**
      * scans the arc and tries to determine the pokemon level, returns 1 if nothing found
      * @param pokemonImage The image of the entire screen
      * @return the estimated pokemon level, or 1 if nothing found
      */
-    private double getPokemonLevel(Bitmap pokemonImage){
+    private double getPokemonLevelFromImg(Bitmap pokemonImage, double trainerLevel){
+        double estimatedPokemonLevel = trainerLevel + 1.5;
         for (double estPokemonLevel = estimatedPokemonLevel; estPokemonLevel >= 1.0; estPokemonLevel -= 0.5) {
             //double angleInDegrees = (Data.CpM[(int) (estPokemonLevel * 2 - 2)] - 0.094) * 202.037116 / Data.CpM[trainerLevel * 2 - 2];
             //if (angleInDegrees > 1.0 && trainerLevel < 30) {
@@ -688,7 +689,7 @@ public class MainActivity extends AppCompatActivity {
     /**
      * get the cp of a pokemon image
      * @param pokemonImage the image of the whole pokemon screen
-     * @return a CP of the pokeon
+     * @return a CP of the pokemon, 10 if scan failed
      */
     private int getPokemonCPFromImg(Bitmap pokemonImage) {
         int pokemonCP = 0;
@@ -719,18 +720,21 @@ public class MainActivity extends AppCompatActivity {
         //WARNING: this method *must* always send an intent at the end, no matter what, to avoid the application hanging.
         Intent info = Pokefly.createNoInfoIntent();
 
-        estimatedPokemonLevel = trainerLevel + 1.5;
-        estimatedPokemonLevel = getPokemonLevel(pokemonImage);
+        try {
+            double estimatedPokemonLevel = getPokemonLevelFromImg(pokemonImage, trainerLevel);
+            String pokemonName = getPokemonNameFromImg(pokemonImage);
+            String candyName = getCandyNameFromImg(pokemonImage);
+            int pokemonHP = getPokemonHPFromImg(pokemonImage);
+            int pokemonCP = getPokemonCPFromImg(pokemonImage);
+            if (candyName.equals("") && pokemonHP == 10 && pokemonCP == 10){ //the default values for a failed scan, if all three fail, then probably scrolled down.
+                Toast.makeText(MainActivity.this, "Please scroll up and scan again", Toast.LENGTH_SHORT).show();
+            }
+            Pokefly.populateInfoIntent(info, pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
+        }finally{
 
-        String pokemonName = getPokemonNameFromImg(pokemonImage);
-        String candyName = getCandyNameFromImg(pokemonImage);
-        int pokemonHP = getPokemonHPFromImg(pokemonImage);
-        int pokemonCP = getPokemonCPFromImg(pokemonImage);
-
-        Pokefly.populateInfoIntent(info, pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
-        LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
+            LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
+        }
     }
-
     /**
      * Dont missgender the poor nidorans.
      *

--- a/app/src/main/java/com/kamron/pogoiv/MainActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/MainActivity.java
@@ -631,84 +631,85 @@ public class MainActivity extends AppCompatActivity {
      */
     private void scanPokemon(Bitmap pokemonImage, String filePath) {
         //WARNING: this method *must* always send an intent at the end, no matter what, to avoid the application hanging.
-        estimatedPokemonLevel = trainerLevel + 1.5;
-
-        estimatedPokemonLevel= getPokemonLevel(pokemonImage);
-
-        Bitmap name = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 4, (int) Math.round(displayMetrics.heightPixels / 2.22608696), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 18.2857143));
-        name = replaceColors(name, 68, 105, 108, Color.WHITE, 200);
-        tesseract.setImage(name);
-        //System.out.println(tesseract.getUTF8Text());
-        pokemonName = tesseract.getUTF8Text().replace(" ", "").replace("1", "l").replace("0", "o").replace("Sparky", getString(R.string.pokemon133)).replace("Rainer", getString(R.string.pokemon133)).replace("Pyro", getString(R.string.pokemon133));
-
-        if (pokemonName.toLowerCase().contains("nidora")){
-            boolean isFemale = isNidoranFemale(pokemonImage);
-            if(isFemale){
-                pokemonName = getResources().getString(R.string.pokemon029);
-            }else{
-                pokemonName = getResources().getString(R.string.pokemon032);
-            }
-        }
-        //SaveImage(name, "name");
-        // TODO : Check rectangle and color
-        Bitmap candy = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 2, (int) Math.round(displayMetrics.heightPixels / 1.3724285), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 38.4));
-        candy = replaceColors(candy, 68, 105, 108, Color.WHITE, 200);
-        tesseract.setImage(candy);
-        //System.out.println(tesseract.getUTF8Text());
-        //SaveImage(candy, "candy");
+        Intent info = Pokefly.createNoInfoIntent();
         try {
-            candyName = tesseract.getUTF8Text().trim().replace("-", " ").split(" ")[candyOrder].replace(" ", "").replace("1", "l").replace("0", "o");
-            candyName = new StringBuilder().append(candyName.substring(0, 1)).append(candyName.substring(1).toLowerCase()).toString();
-        }catch(StringIndexOutOfBoundsException e){
-            candyName = pokemonName; //Default for not finding candy name
-        }
-        Bitmap hp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 2.8), (int) Math.round(displayMetrics.heightPixels / 1.8962963), (int) Math.round(displayMetrics.widthPixels / 3.5), (int) Math.round(displayMetrics.heightPixels / 34.13333333));
-        hp = replaceColors(hp, 55, 66, 61, Color.WHITE, 200);
-        tesseract.setImage(hp);
-        //System.out.println(tesseract.getUTF8Text());
-        String pokemonHPStr = tesseract.getUTF8Text();
+            estimatedPokemonLevel = trainerLevel + 1.5;
 
-        //Check if valid pokemon TODO find a better method of determining whether or not this is a pokemon image
-        if(pokemonHPStr.contains("/")) {
+            estimatedPokemonLevel = getPokemonLevel(pokemonImage);
+
+            Bitmap name = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 4, (int) Math.round(displayMetrics.heightPixels / 2.22608696), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 18.2857143));
+            name = replaceColors(name, 68, 105, 108, Color.WHITE, 200);
+            tesseract.setImage(name);
+            //System.out.println(tesseract.getUTF8Text());
+            pokemonName = tesseract.getUTF8Text().replace(" ", "").replace("1", "l").replace("0", "o").replace("Sparky", getString(R.string.pokemon133)).replace("Rainer", getString(R.string.pokemon133)).replace("Pyro", getString(R.string.pokemon133));
+
+            if (pokemonName.toLowerCase().contains("nidora")) {
+                boolean isFemale = isNidoranFemale(pokemonImage);
+                if (isFemale) {
+                    pokemonName = getResources().getString(R.string.pokemon029);
+                } else {
+                    pokemonName = getResources().getString(R.string.pokemon032);
+                }
+            }
+            //SaveImage(name, "name");
+            // TODO : Check rectangle and color
+            Bitmap candy = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 2, (int) Math.round(displayMetrics.heightPixels / 1.3724285), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 38.4));
+            candy = replaceColors(candy, 68, 105, 108, Color.WHITE, 200);
+            tesseract.setImage(candy);
+            //System.out.println(tesseract.getUTF8Text());
+            //SaveImage(candy, "candy");
             try {
-                pokemonHP = Integer.parseInt(pokemonHPStr.split("/")[1].replace("Z", "2").replace("O", "0").replace("l", "1").replaceAll("[^0-9]", ""));
-            } catch (java.lang.NumberFormatException e) {
-                pokemonHP = 10;
+                candyName = tesseract.getUTF8Text().trim().replace("-", " ").split(" ")[candyOrder].replace(" ", "").replace("1", "l").replace("0", "o");
+                candyName = new StringBuilder().append(candyName.substring(0, 1)).append(candyName.substring(1).toLowerCase()).toString();
+            } catch (StringIndexOutOfBoundsException e) {
+                candyName = pokemonName; //Default for not finding candy name
             }
+            Bitmap hp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 2.8), (int) Math.round(displayMetrics.heightPixels / 1.8962963), (int) Math.round(displayMetrics.widthPixels / 3.5), (int) Math.round(displayMetrics.heightPixels / 34.13333333));
+            hp = replaceColors(hp, 55, 66, 61, Color.WHITE, 200);
+            tesseract.setImage(hp);
+            //System.out.println(tesseract.getUTF8Text());
+            String pokemonHPStr = tesseract.getUTF8Text();
 
-            //SaveImage(hp, "hp");
-            Bitmap cp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 3.0), (int) Math.round(displayMetrics.heightPixels / 15.5151515), (int) Math.round(displayMetrics.widthPixels / 3.84), (int) Math.round(displayMetrics.heightPixels / 21.333333333));
-            cp = replaceColors(cp, 255, 255, 255, Color.BLACK, 30);
-            tesseract.setImage(cp);
-            //String cpText = tesseract.getUTF8Text().replace("O", "0").replace("l", "1").replace("S", "3").replaceAll("[^0-9]", "");
-            String cpText = tesseract.getUTF8Text().replace("O", "0").replace("l", "1");
-            if (cpText.length() >=2){ //gastly can block the "cp" text, so its not visible...
-                cpText = cpText.substring(2);
-            }
-            //System.out.println(cpText);
-            try {
-                pokemonCP = Integer.parseInt(cpText);
-            } catch (java.lang.NumberFormatException e) {
-                pokemonCP = 10;
-            }
+            //Check if valid pokemon TODO find a better method of determining whether or not this is a pokemon image
+            if (pokemonHPStr.contains("/")) {
+                try {
+                    pokemonHP = Integer.parseInt(pokemonHPStr.split("/")[1].replace("Z", "2").replace("O", "0").replace("l", "1").replaceAll("[^0-9]", ""));
+                } catch (java.lang.NumberFormatException e) {
+                    pokemonHP = 10;
+                }
 
-            if (pokemonCP > 4500) {
-                cpText = cpText.substring(1);
-                pokemonCP = Integer.parseInt(cpText);
-            }
-            //SaveImage(cp, "cp");
-            //System.out.println("Name: " + pokemonName);
-            //System.out.println("HP: " + pokemonHP);
-            //System.out.println("CP: " + pokemonCP);
-            name.recycle();
-            candy.recycle();
-            cp.recycle();
-            hp.recycle();
+                //SaveImage(hp, "hp");
+                Bitmap cp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 3.0), (int) Math.round(displayMetrics.heightPixels / 15.5151515), (int) Math.round(displayMetrics.widthPixels / 3.84), (int) Math.round(displayMetrics.heightPixels / 21.333333333));
+                cp = replaceColors(cp, 255, 255, 255, Color.BLACK, 30);
+                tesseract.setImage(cp);
+                //String cpText = tesseract.getUTF8Text().replace("O", "0").replace("l", "1").replace("S", "3").replaceAll("[^0-9]", "");
+                String cpText = tesseract.getUTF8Text().replace("O", "0").replace("l", "1");
+                if (cpText.length() >= 2) { //gastly can block the "cp" text, so its not visible...
+                    cpText = cpText.substring(2);
+                }
+                //System.out.println(cpText);
+                try {
+                    pokemonCP = Integer.parseInt(cpText);
+                } catch (java.lang.NumberFormatException e) {
+                    pokemonCP = 10;
+                }
 
-            Intent info = Pokefly.createInfoIntent(pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
-            LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
-        } else {
-            Intent info = Pokefly.createNoInfoIntent();
+                if (pokemonCP > 4500) {
+                    cpText = cpText.substring(1);
+                    pokemonCP = Integer.parseInt(cpText);
+                }
+                //SaveImage(cp, "cp");
+                //System.out.println("Name: " + pokemonName);
+                //System.out.println("HP: " + pokemonHP);
+                //System.out.println("CP: " + pokemonCP);
+                name.recycle();
+                candy.recycle();
+                cp.recycle();
+                hp.recycle();
+
+                Pokefly.populateInfoIntent(info, pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
+            }
+        } finally {
             LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
         }
     }

--- a/app/src/main/java/com/kamron/pogoiv/MainActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/MainActivity.java
@@ -708,6 +708,8 @@ public class MainActivity extends AppCompatActivity {
                 hp.recycle();
 
                 Pokefly.populateInfoIntent(info, pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
+            } else {
+                Toast.makeText(this, R.string.scan_pokemon_failed, Toast.LENGTH_SHORT).show();
             }
         } finally {
             LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);

--- a/app/src/main/java/com/kamron/pogoiv/MainActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/MainActivity.java
@@ -39,7 +39,6 @@ import android.support.v7.app.AlertDialog;
 import android.support.v7.app.AppCompatActivity;
 import android.text.method.LinkMovementMethod;
 import android.util.DisplayMetrics;
-import android.util.Log;
 import android.view.Display;
 import android.view.Menu;
 import android.view.MenuInflater;
@@ -109,12 +108,8 @@ public class MainActivity extends AppCompatActivity {
 
     private boolean readyForNewScreenshot = true;
 
-    private String pokemonName;
-    private String candyName;
     private int candyOrder = 0;
     private double estimatedPokemonLevel;
-    private int pokemonCP;
-    private int pokemonHP;
     private boolean pokeFlyRunning = false;
     private int trainerLevel;
 
@@ -622,6 +617,97 @@ public class MainActivity extends AppCompatActivity {
         }
         return 1;
     }
+
+    /**
+     * get the pokemon name as analysed from a pokemon image
+     * @param pokemonImage the image of the whole screen
+     * @return A string resulting from the scan
+     */
+    private String getPokemonNameFromImg(Bitmap pokemonImage){
+        Bitmap name = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 4, (int) Math.round(displayMetrics.heightPixels / 2.22608696), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 18.2857143));
+        name = replaceColors(name, 68, 105, 108, Color.WHITE, 200);
+        tesseract.setImage(name);
+        //System.out.println(tesseract.getUTF8Text());
+        String pokemonName = tesseract.getUTF8Text().replace(" ", "").replace("1", "l").replace("0", "o").replace("Sparky", getString(R.string.pokemon133)).replace("Rainer", getString(R.string.pokemon133)).replace("Pyro", getString(R.string.pokemon133));
+        if (pokemonName.toLowerCase().contains("nidora")) {
+            boolean isFemale = isNidoranFemale(pokemonImage);
+            if (isFemale) {
+                pokemonName = getResources().getString(R.string.pokemon029);
+            } else {
+                pokemonName = getResources().getString(R.string.pokemon032);
+            }
+        }
+        name.recycle();
+        return pokemonName;
+    }
+
+    /**
+     * gets the candy name from a pokenon image
+     * @param pokemonImage the image of the whole screen
+     * @return the candy name, or "" if nothing was found
+     */
+    private String getCandyNameFromImg(Bitmap pokemonImage){
+        Bitmap candy = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 2, (int) Math.round(displayMetrics.heightPixels / 1.3724285), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 38.4));
+        candy = replaceColors(candy, 68, 105, 108, Color.WHITE, 200);
+        tesseract.setImage(candy);
+        String candyName ="";
+        //System.out.println(tesseract.getUTF8Text());
+        //SaveImage(candy, "candy");
+        try {
+            candyName = tesseract.getUTF8Text().trim().replace("-", " ").split(" ")[candyOrder].replace(" ", "").replace("1", "l").replace("0", "o");
+            candyName = new StringBuilder().append(candyName.substring(0, 1)).append(candyName.substring(1).toLowerCase()).toString();
+        } catch (StringIndexOutOfBoundsException e) {
+            candyName = "";
+        }
+        candy.recycle();
+        return candyName;
+    }
+
+    /**
+     * get the pokemon hp from a picture
+     * @param pokemonImage the image of the whole screen
+     * @return an integer of the interpreted pokemon name, 10 if scan failed
+     */
+    private int getPokemonHPFromImg(Bitmap pokemonImage){
+        int pokemonHP = 10;
+        Bitmap hp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 2.8), (int) Math.round(displayMetrics.heightPixels / 1.8962963), (int) Math.round(displayMetrics.widthPixels / 3.5), (int) Math.round(displayMetrics.heightPixels / 34.13333333));
+        hp = replaceColors(hp, 55, 66, 61, Color.WHITE, 200);
+        tesseract.setImage(hp);
+        String pokemonHPStr = tesseract.getUTF8Text();
+        if (pokemonHPStr.contains("/")) {
+            try {
+                pokemonHP = Integer.parseInt(pokemonHPStr.split("/")[1].replace("Z", "2").replace("O", "0").replace("l", "1").replaceAll("[^0-9]", ""));
+            } catch (java.lang.NumberFormatException e) {
+                pokemonHP = 10;
+            }
+        }
+        hp.recycle();
+        return pokemonHP;
+    }
+
+    /**
+     * get the cp of a pokemon image
+     * @param pokemonImage the image of the whole pokemon screen
+     * @return a CP of the pokeon
+     */
+    private int getPokemonCPFromImg(Bitmap pokemonImage) {
+        int pokemonCP = 0;
+        Bitmap cp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 3.0), (int) Math.round(displayMetrics.heightPixels / 15.5151515), (int) Math.round(displayMetrics.widthPixels / 3.84), (int) Math.round(displayMetrics.heightPixels / 21.333333333));
+        cp = replaceColors(cp, 255, 255, 255, Color.BLACK, 30);
+        tesseract.setImage(cp);
+        String cpText = tesseract.getUTF8Text().replace("O", "0").replace("l", "1");
+        if (cpText.length() >= 2) { //gastly can block the "cp" text, so its not visible...
+            cpText = cpText.substring(2);
+        }
+        try {
+            pokemonCP = Integer.parseInt(cpText);
+        } catch (java.lang.NumberFormatException e) {
+            pokemonCP = 10;
+        }
+        cp.recycle();
+        return  pokemonCP;
+    }
+
     /**
      * scanPokemon
      * Performs OCR on an image of a pokemon and sends the pulled info to PokeFly to display.
@@ -632,88 +718,17 @@ public class MainActivity extends AppCompatActivity {
     private void scanPokemon(Bitmap pokemonImage, String filePath) {
         //WARNING: this method *must* always send an intent at the end, no matter what, to avoid the application hanging.
         Intent info = Pokefly.createNoInfoIntent();
-        try {
-            estimatedPokemonLevel = trainerLevel + 1.5;
 
-            estimatedPokemonLevel = getPokemonLevel(pokemonImage);
+        estimatedPokemonLevel = trainerLevel + 1.5;
+        estimatedPokemonLevel = getPokemonLevel(pokemonImage);
 
-            Bitmap name = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 4, (int) Math.round(displayMetrics.heightPixels / 2.22608696), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 18.2857143));
-            name = replaceColors(name, 68, 105, 108, Color.WHITE, 200);
-            tesseract.setImage(name);
-            //System.out.println(tesseract.getUTF8Text());
-            pokemonName = tesseract.getUTF8Text().replace(" ", "").replace("1", "l").replace("0", "o").replace("Sparky", getString(R.string.pokemon133)).replace("Rainer", getString(R.string.pokemon133)).replace("Pyro", getString(R.string.pokemon133));
+        String pokemonName = getPokemonNameFromImg(pokemonImage);
+        String candyName = getCandyNameFromImg(pokemonImage);
+        int pokemonHP = getPokemonHPFromImg(pokemonImage);
+        int pokemonCP = getPokemonCPFromImg(pokemonImage);
 
-            if (pokemonName.toLowerCase().contains("nidora")) {
-                boolean isFemale = isNidoranFemale(pokemonImage);
-                if (isFemale) {
-                    pokemonName = getResources().getString(R.string.pokemon029);
-                } else {
-                    pokemonName = getResources().getString(R.string.pokemon032);
-                }
-            }
-            //SaveImage(name, "name");
-            // TODO : Check rectangle and color
-            Bitmap candy = Bitmap.createBitmap(pokemonImage, displayMetrics.widthPixels / 2, (int) Math.round(displayMetrics.heightPixels / 1.3724285), (int) Math.round(displayMetrics.widthPixels / 2.057), (int) Math.round(displayMetrics.heightPixels / 38.4));
-            candy = replaceColors(candy, 68, 105, 108, Color.WHITE, 200);
-            tesseract.setImage(candy);
-            //System.out.println(tesseract.getUTF8Text());
-            //SaveImage(candy, "candy");
-            try {
-                candyName = tesseract.getUTF8Text().trim().replace("-", " ").split(" ")[candyOrder].replace(" ", "").replace("1", "l").replace("0", "o");
-                candyName = new StringBuilder().append(candyName.substring(0, 1)).append(candyName.substring(1).toLowerCase()).toString();
-            } catch (StringIndexOutOfBoundsException e) {
-                candyName = pokemonName; //Default for not finding candy name
-            }
-            Bitmap hp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 2.8), (int) Math.round(displayMetrics.heightPixels / 1.8962963), (int) Math.round(displayMetrics.widthPixels / 3.5), (int) Math.round(displayMetrics.heightPixels / 34.13333333));
-            hp = replaceColors(hp, 55, 66, 61, Color.WHITE, 200);
-            tesseract.setImage(hp);
-            //System.out.println(tesseract.getUTF8Text());
-            String pokemonHPStr = tesseract.getUTF8Text();
-
-            //Check if valid pokemon TODO find a better method of determining whether or not this is a pokemon image
-            if (pokemonHPStr.contains("/")) {
-                try {
-                    pokemonHP = Integer.parseInt(pokemonHPStr.split("/")[1].replace("Z", "2").replace("O", "0").replace("l", "1").replaceAll("[^0-9]", ""));
-                } catch (java.lang.NumberFormatException e) {
-                    pokemonHP = 10;
-                }
-
-                //SaveImage(hp, "hp");
-                Bitmap cp = Bitmap.createBitmap(pokemonImage, (int) Math.round(displayMetrics.widthPixels / 3.0), (int) Math.round(displayMetrics.heightPixels / 15.5151515), (int) Math.round(displayMetrics.widthPixels / 3.84), (int) Math.round(displayMetrics.heightPixels / 21.333333333));
-                cp = replaceColors(cp, 255, 255, 255, Color.BLACK, 30);
-                tesseract.setImage(cp);
-                //String cpText = tesseract.getUTF8Text().replace("O", "0").replace("l", "1").replace("S", "3").replaceAll("[^0-9]", "");
-                String cpText = tesseract.getUTF8Text().replace("O", "0").replace("l", "1");
-                if (cpText.length() >= 2) { //gastly can block the "cp" text, so its not visible...
-                    cpText = cpText.substring(2);
-                }
-                //System.out.println(cpText);
-                try {
-                    pokemonCP = Integer.parseInt(cpText);
-                } catch (java.lang.NumberFormatException e) {
-                    pokemonCP = 10;
-                }
-
-                if (pokemonCP > 4500) {
-                    cpText = cpText.substring(1);
-                    pokemonCP = Integer.parseInt(cpText);
-                }
-                //SaveImage(cp, "cp");
-                //System.out.println("Name: " + pokemonName);
-                //System.out.println("HP: " + pokemonHP);
-                //System.out.println("CP: " + pokemonCP);
-                name.recycle();
-                candy.recycle();
-                cp.recycle();
-                hp.recycle();
-
-                Pokefly.populateInfoIntent(info, pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
-            } else {
-                Toast.makeText(this, R.string.scan_pokemon_failed, Toast.LENGTH_SHORT).show();
-            }
-        } finally {
-            LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
-        }
+        Pokefly.populateInfoIntent(info, pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
+        LocalBroadcastManager.getInstance(MainActivity.this).sendBroadcast(info);
     }
 
     /**

--- a/app/src/main/java/com/kamron/pogoiv/MainActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/MainActivity.java
@@ -727,7 +727,7 @@ public class MainActivity extends AppCompatActivity {
             int pokemonHP = getPokemonHPFromImg(pokemonImage);
             int pokemonCP = getPokemonCPFromImg(pokemonImage);
             if (candyName.equals("") && pokemonHP == 10 && pokemonCP == 10){ //the default values for a failed scan, if all three fail, then probably scrolled down.
-                Toast.makeText(MainActivity.this, "Please scroll up and scan again", Toast.LENGTH_SHORT).show();
+                Toast.makeText(MainActivity.this, getString(R.string.scan_pokemon_failed), Toast.LENGTH_SHORT).show();
             }
             Pokefly.populateInfoIntent(info, pokemonName, candyName, pokemonHP, pokemonCP, estimatedPokemonLevel, filePath);
         }finally{

--- a/app/src/main/java/com/kamron/pogoiv/Pokefly.java
+++ b/app/src/main/java/com/kamron/pogoiv/Pokefly.java
@@ -158,9 +158,7 @@ public class Pokefly extends Service {
     public static Intent createNoInfoIntent() {
         return new Intent(ACTION_SEND_INFO);
     }
-
-    public static Intent createInfoIntent(String pokemonName, String candyName, int pokemonHP, int pokemonCP, double estimatedPokemonLevel, String filePath) {
-        Intent intent = createNoInfoIntent();
+    public static void populateInfoIntent(Intent intent, String pokemonName, String candyName, int pokemonHP, int pokemonCP, double estimatedPokemonLevel, String filePath) {
         intent.putExtra(KEY_SEND_INFO_NAME, pokemonName);
         intent.putExtra(KEY_SEND_INFO_CANDY, candyName);
         intent.putExtra(KEY_SEND_INFO_HP, pokemonHP);
@@ -169,7 +167,6 @@ public class Pokefly extends Service {
         if (!filePath.isEmpty()) {
             intent.putExtra(KEY_SEND_SCREENSHOT_DIR, filePath);
         }
-        return intent;
     }
 
     @Override

--- a/app/src/main/java/com/kamron/pogoiv/Pokefly.java
+++ b/app/src/main/java/com/kamron/pogoiv/Pokefly.java
@@ -27,8 +27,6 @@ import android.view.LayoutInflater;
 import android.view.MotionEvent;
 import android.view.View;
 import android.view.WindowManager;
-import android.widget.ArrayAdapter;
-import android.widget.Button;
 import android.widget.EditText;
 import android.widget.ImageView;
 import android.widget.LinearLayout;
@@ -82,7 +80,7 @@ public class Pokefly extends Service {
     private DisplayMetrics displayMetrics;
     ClipboardManager clipboard;
 
-    private boolean infoShown = false;
+    private boolean infoShownSent = false;
     private boolean infoShownReceived = false;
     private boolean IVButtonShown = false;
 
@@ -367,7 +365,7 @@ public class Pokefly extends Service {
                         Intent intent = MainActivity.createScreenshotIntent();
                         LocalBroadcastManager.getInstance(Pokefly.this).sendBroadcast(intent);
                         receivedInfo = false;
-                        infoShown = true;
+                        infoShownSent = true;
                         infoShownReceived = false;
                         break;
                 }
@@ -436,7 +434,7 @@ public class Pokefly extends Service {
             IVButtonShown = true;
         }
         receivedInfo = false;
-        infoShown = false;
+        infoShownSent = false;
         Intent resetIntent = MainActivity.createResetScreenshotIntent();
         LocalBroadcastManager.getInstance(Pokefly.this).sendBroadcast(resetIntent);
     }
@@ -730,7 +728,7 @@ public class Pokefly extends Service {
         @Override
         public void onReceive(Context context, Intent intent) {
             boolean show = intent.getBooleanExtra(KEY_DISPLAY_IV_BUTTON_SHOW, false);
-            if (show && !IVButtonShown && !infoShown) {
+            if (show && !IVButtonShown && !infoShownSent) {
                 windowManager.addView(IVButton, IVButonParams);
                 IVButtonShown = true;
             } else if (!show) {

--- a/app/src/main/java/com/kamron/pogoiv/Pokefly.java
+++ b/app/src/main/java/com/kamron/pogoiv/Pokefly.java
@@ -214,16 +214,28 @@ public class Pokefly extends Service {
         return START_STICKY;
     }
 
+    private boolean infoLayoutArcPointerVisible = false;
+    private void showInfoLayoutArcPointer() {
+        if (!infoLayoutArcPointerVisible && arcPointer != null && infoLayout != null) {
+            infoLayoutArcPointerVisible = true;
+            windowManager.addView(arcPointer, arcParams);
+            windowManager.addView(infoLayout, layoutParams);
+        }
+    }
+
+    private void hideInfoLayoutArcPointer() {
+        if (infoLayoutArcPointerVisible) {
+            windowManager.removeView(arcPointer);
+            windowManager.removeView(infoLayout);
+            infoLayoutArcPointerVisible = false;
+        }
+    }
 
     @Override
     public void onDestroy() {
         super.onDestroy();
         if (IVButton != null && IVButtonShown) windowManager.removeView(IVButton);
-        if (infoShown) {
-            if (arcPointer != null) windowManager.removeView(arcPointer);
-            //if(arcAdjustBar != null) windowManager.removeView(arcAdjustBar);
-            if (infoLayout != null) windowManager.removeView(infoLayout);
-        }
+        hideInfoLayoutArcPointer();
         stopForeground(true);
         LocalBroadcastManager.getInstance(this).unregisterReceiver(displayInfo);
         LocalBroadcastManager.getInstance(this).unregisterReceiver(setIVButtonDisplay);
@@ -427,8 +439,7 @@ public class Pokefly extends Service {
 
     @OnClick({ R.id.btnCancelInfo, R.id.btnCloseInfo })
     public void cancelInfoDialog() {
-        windowManager.removeView(infoLayout);
-        windowManager.removeView(arcPointer);
+        hideInfoLayoutArcPointer();
         if(!batterySaver) {
             windowManager.addView(IVButton, IVButonParams);
             IVButtonShown = true;
@@ -477,8 +488,7 @@ public class Pokefly extends Service {
             pokemonHPEdit.setText(String.valueOf(pokemonHP));
             pokemonCPEdit.setText(String.valueOf(pokemonCP));
 
-            windowManager.addView(arcPointer, arcParams);
-            windowManager.addView(infoLayout, layoutParams);
+            showInfoLayoutArcPointer();
             setArcPointer(estimatedPokemonLevel);
             //setArcPointer((Data.CpM[(int) (estimatedPokemonLevel * 2 - 2)] - 0.094) * 202.037116 / Data.CpM[trainerLevel * 2 - 2]);
             arcAdjustBar.setProgress((int) ((estimatedPokemonLevel - 1) * 2));

--- a/app/src/main/java/com/kamron/pogoiv/Pokefly.java
+++ b/app/src/main/java/com/kamron/pogoiv/Pokefly.java
@@ -155,8 +155,12 @@ public class Pokefly extends Service {
         return intent;
     }
 
+    public static Intent createNoInfoIntent() {
+        return new Intent(ACTION_SEND_INFO);
+    }
+
     public static Intent createInfoIntent(String pokemonName, String candyName, int pokemonHP, int pokemonCP, double estimatedPokemonLevel, String filePath) {
-        Intent intent = new Intent(ACTION_SEND_INFO);
+        Intent intent = createNoInfoIntent();
         intent.putExtra(KEY_SEND_INFO_NAME, pokemonName);
         intent.putExtra(KEY_SEND_INFO_CANDY, candyName);
         intent.putExtra(KEY_SEND_INFO_HP, pokemonHP);
@@ -444,6 +448,13 @@ public class Pokefly extends Service {
             windowManager.addView(IVButton, IVButonParams);
             IVButtonShown = true;
         }
+        resetPokeflyStateMachine();
+    }
+
+    /**
+     * Reset service state so that a new pokemon info can be requested.
+     */
+    private void resetPokeflyStateMachine() {
         receivedInfo = false;
         infoShownSent = false;
         Intent resetIntent = MainActivity.createResetScreenshotIntent();
@@ -708,23 +719,27 @@ public class Pokefly extends Service {
     private BroadcastReceiver displayInfo = new BroadcastReceiver() {
         @Override
         public void onReceive(Context context, Intent intent) {
-            if (!receivedInfo && intent.hasExtra(KEY_SEND_INFO_NAME) && intent.hasExtra(KEY_SEND_INFO_CP) && intent.hasExtra(KEY_SEND_INFO_HP) && intent.hasExtra(KEY_SEND_INFO_LEVEL)) {
+            if (!receivedInfo) {
                 receivedInfo = true;
-                pokemonName = intent.getStringExtra(KEY_SEND_INFO_NAME);
-                candyName = intent.getStringExtra(KEY_SEND_INFO_CANDY);
-                pokemonCP = intent.getIntExtra(KEY_SEND_INFO_CP, 0);
-                pokemonHP = intent.getIntExtra(KEY_SEND_INFO_HP, 0);
-                estimatedPokemonLevel = intent.getDoubleExtra(KEY_SEND_INFO_LEVEL, estimatedPokemonLevel);
-                if (estimatedPokemonLevel < 1.0) {
-                    estimatedPokemonLevel = 1.0;
-                }
-                if (intent.hasExtra(KEY_SEND_SCREENSHOT_DIR)) {
-                    screenshotDir = intent.getStringExtra(KEY_SEND_SCREENSHOT_DIR);
-                } else {
-                    screenshotDir = "";
-                }
+                if (intent.hasExtra(KEY_SEND_INFO_NAME) && intent.hasExtra(KEY_SEND_INFO_CP) && intent.hasExtra(KEY_SEND_INFO_HP) && intent.hasExtra(KEY_SEND_INFO_LEVEL)) {
+                    pokemonName = intent.getStringExtra(KEY_SEND_INFO_NAME);
+                    candyName = intent.getStringExtra(KEY_SEND_INFO_CANDY);
+                    pokemonCP = intent.getIntExtra(KEY_SEND_INFO_CP, 0);
+                    pokemonHP = intent.getIntExtra(KEY_SEND_INFO_HP, 0);
+                    estimatedPokemonLevel = intent.getDoubleExtra(KEY_SEND_INFO_LEVEL, estimatedPokemonLevel);
+                    if (estimatedPokemonLevel < 1.0) {
+                        estimatedPokemonLevel = 1.0;
+                    }
+                    if (intent.hasExtra(KEY_SEND_SCREENSHOT_DIR)) {
+                        screenshotDir = intent.getStringExtra(KEY_SEND_SCREENSHOT_DIR);
+                    } else {
+                        screenshotDir = "";
+                    }
 
-                showInfoLayout();
+                    showInfoLayout();
+                } else {
+                    resetPokeflyStateMachine();
+                }
             }
         }
     };

--- a/app/src/main/java/com/kamron/pogoiv/ShareHandlerActivity.java
+++ b/app/src/main/java/com/kamron/pogoiv/ShareHandlerActivity.java
@@ -26,8 +26,7 @@ public class ShareHandlerActivity extends Activity {
             Uri imageUri = (Uri) intent.getParcelableExtra(Intent.EXTRA_STREAM);
             try {
                 Bitmap bitmap = MediaStore.Images.Media.getBitmap(this.getContentResolver(), imageUri);
-                Intent newintent = new Intent("process-bitmap");
-                newintent.putExtra("bitmap", bitmap);
+                Intent newintent = MainActivity.createProcessBitmapIntent(bitmap);
                 LocalBroadcastManager.getInstance(ShareHandlerActivity.this).sendBroadcast(newintent);
             } catch (IOException e) {
                 e.printStackTrace();

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -215,5 +215,5 @@
     <string name="auto_update_setting_summary">Automatically check for updates when the app starts up</string>
     <string name="check_for_update_setting_title">Check for update</string>
     <string name="check_for_update_setting_summary">See if an update is available</string>
-    <string name="scan_pokemon_failed">Scanning Pokémon failed, scroll to the top and try again!</string>
+    <string name="scan_pokemon_failed">Scanning failed, please scroll up and try again!</string>
 </resources>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -180,7 +180,7 @@
     <string name="ivtext_iv">Min: %1$s%% Average: %2$s%% Max: %3$s%%\n</string>
     <string name="ivtext_evolve">Can be evolved into %1$s:\n</string>
     <string name="ivtext_evolve_further">Can be further evolved into %1$s:</string>
-    <string name="ivtext_many_possibilities">There are too many possibilities for this pokemon. Try powering it up!</string>
+    <string name="ivtext_many_possibilities">There are too many possibilities for this Pokémon. Try powering it up!</string>
     <string name="ivtext_max_lvl_cost">Cost to reach lvl %1$s:</string>
     <string name="ivtext_max_lvl_cost2">Candy: %1$s Stardust: %2$s</string>
     <string name="main_battery_saver">Battery Saver (Screenshot to Scan)</string>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -215,4 +215,5 @@
     <string name="auto_update_setting_summary">Automatically check for updates when the app starts up</string>
     <string name="check_for_update_setting_title">Check for update</string>
     <string name="check_for_update_setting_summary">See if an update is available</string>
+    <string name="scan_pokemon_failed">Scanning Pokémon failed, scroll to the top and try again!</string>
 </resources>

